init commit

HarlamovBuldog opened the pull request.

HarlamovBuldog committed:
init commit

pisarukv commented on main.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+func main() {
+	wget.Execute()
```
better put all inside main

pisarukv commented on wget/Example_test.go:
```diff
@@ -0,0 +1,56 @@
+	}
+}
+
+func WasFileDownloaded(link string, totalSizeURL int64) (bool, error) {
```
why a u exporting the function?

pisarukv replied:
And of course isFileDownloaded, but more go way it to name it fileDownloaded

pisarukv replied:
pls check https://golang.org/doc/effective_go.html

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+
+type DownloadManager struct {
+	fileList        map[string]*File // key = link
+	fileListOrder   []string         // to order fileList map iteration
```
fileList can be slice, and you can do it without fileListOrder

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+	}
+}
+
+func ParseURL(link string) error {
```
What actually dose this function dose? ParseURL but why it returns only error? Need better naming here

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		return fmt.Errorf("wget: content length = %v :%v", bodySize, err)
+		//fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
```
delete dead code always

HarlamovBuldog replied:
Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+	var linksList []string
+	dm := NewDownloadManager()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
```
ctx, cancel := context.WithCancel(context.Background())

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+func Execute() {
+	wg := &sync.WaitGroup{}
+	var linksList []string
+	dm := NewDownloadManager()
```
create vars just before the place where are they used

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+		}
+	}
+
+	// locking map here
```
dm := NewDownloadManager()

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+		dm.fileListLock.Lock()
+		for _, key := range dm.fileListOrder {
+			if dm.fileList[key].totalSize != 0 {
+				fmt.Printf("%.2f %% ", float64(dm.fileList[key].counter)/float64(dm.fileList[key].totalSize)*100)
```
put into a separate var

cali4888 commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+		case <-ctx.Done():
+			print()
+			return
+		default:
```
it's possible here to write case <- time.NewTicker(100 * time.Millisecond):

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error getting content length :%v", err)
```
dead code

HarlamovBuldog replied:
Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+	defer wg.Done()
+	resp, err := http.Get(link)
+	if err != nil {
+		//return false, 0, fmt.Errorf("wget: error connecting url :%v", err)
```
dead code

HarlamovBuldog replied:
Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+		return
+	}
+
+	//return true, int(bytes), nil
```
dead code

HarlamovBuldog replied:
Have deleted dead code in commit 42bef00c07d5f5b947db74850f8533ee10ed0aa0

pisarukv commented on wget/wget.go:
```diff
@@ -0,0 +1,202 @@
+	// Copy need to be changed to Tee.Reader
+	_, err1 := io.Copy(f, file)
+	if err1 != nil {
+		//return false, 0, fmt.Errorf("wget: error reading from %s: %v", link, err)
```
dead code

pisarukv commented on wget/Example_test.go:
```diff
@@ -0,0 +1,56 @@
+package wget
```
Example_test.go -> wget_test.go

pisarukv commented on main.go:
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"./wget"
```
in go it's common to use absolut imports

pisarukv replied:
so in your case it would be `gitHub.com/HarlamovBuldog/wget-golang-util/wget`

HarlamovBuldog committed:
Dead code was deleted

HarlamovBuldog committed:
Update data structure and parseURL method

DownloadManager structure was deleted in order to simplify solution.
Add comments to parseURL method.
This method returns string and error now.

HarlamovBuldog committed:
Update Download method and delete some dead code

Download is now the method of type File.
No need to lock map.
We just change the value of passed File, which at time of caliing func is already a member of slice []*File.
Also all declarations of values are in right order now, just before actual using of them in code.

HarlamovBuldog committed:
Update Print method

Print method takes 3 arguments now: context, slice type of *File  and channel struct{}.
Again we don't need to lock anything now.
We iterate over slice of Files and using fields of each FIle itself to  calculate percentage of download.
Percentage value was moved to variable.

HarlamovBuldog committed:
Deleted main.go file, implemented main func in wget.go

Main.go file was deleted, cause there is no need to separate files in project and make package with them. Project is just too small.
Execute func in wget.go was reworked in order to work well with new data structure.
Execute func also was renamed to main func, as well as was package wget was renamed to package main.
Also all declarations of variables were relocated place before their actual call.

HarlamovBuldog committed:
Update some comments

HarlamovBuldog committed:
wget.go was renamed to main.go and moved to root folder

HarlamovBuldog committed:
Update Example_test.go

Logic was changed in order to work correctly with new data structure.
WasFileDownloaded func was renamed to fileDownloaded in order to follow golang common practice.
Some variables were created, other were renamed for better code understanding.

HarlamovBuldog committed:
Rename and relocate Example_test.go

Example_test.go was renamed to wget_test.go and relocated to root directory in order to follow common golang writing code practice.

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+// Download tries to download file from remote address using
+// given url. Takes WaitGroup pointer to provide parallelism.
+// Prints error to os.Stderr and return if something goes wrong.
+func (file *File) Download(link string, wg *sync.WaitGroup) {
```
do you want download to be public method?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	var fileList []*File
```
at this moment you already know number of links, so why not fileList := make([]*File, Len(linksList)) ?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+
+	for _, link := range linksList {
+		file := &File{}
+		fileList = append(fileList, file)
```
fileList = append(fileList, &File{})

HarlamovBuldog replied:
But how then i will run file.Download method after append? I mean i need this pointer.

pisarukv replied:
agree

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+		}
+	}
+
+	for {
```
```
	for {
 		select {
 		case <-ctx.Done():
 			print()
 			return
 		default:
 			print()
 			time.Sleep(100 * time.Millisecond)
 		}
 	}
```

```
        t := time.NewTicker(100 * time.Millisecond)
        defer t.Stop()
	for {
 		select {
 		case <-ctx.Done():
 			print()
 			return
 		case <- t.C:
 			print()
 		}
 	}
```

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+	}
+	defer resp.Body.Close()
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
```
so, you already getting and checking content-length in parse url? why you need to do double work?

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+
+	bodySize, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "wget: error getting content length :%v\n", err)
```
log.Printf - will write err to stderr

pisarukv commented on main.go:
```diff
@@ -0,0 +1,175 @@
+	file.totalSize = bodySize
+
+	// Copy need to be changed to Tee.Reader
+	_, err1 := io.Copy(f, file)
```
why err1?

HarlamovBuldog replied:
cause err was declared below. If I will name it err, vscode will tell me smth like "No new variable was declared on the left side"

HarlamovBuldog replied:
forget, i can just type = instead of :=

HarlamovBuldog committed:
Update code structure main.go

Some functions code was relocated in order for better reading and to follow common writing code rules

HarlamovBuldog committed:
Made all functions and File struct unexported

HarlamovBuldog committed:
Update parseURL func in main.go

parseURL func now returns bodySize too in order not to repeat check in Download method.
Return parameters became named.
Comments describing func were updated.

HarlamovBuldog committed:
Update main func in main.go

Downloading and parsing files functionalities now locate in one range.
In this case we don't need links slice anymore.

HarlamovBuldog committed:
Update download func in main.go

We don't need to check content-length and writing it to the file anymore, cause at the moment of calling download func we've done it already in Parse func.

HarlamovBuldog committed:
Added log, updated print func in main.go

All occurrences of fmt.Fprintf were changed to log.Printf
time.NewTicker was added to Print func

HarlamovBuldog committed:
Update wget_test.go

Code was adapted in order to work correctly with updated download method.

pisarukv reviewed (approved):
looks good
